File: xiaohao/数组/350/350_双指针解法.py
# 排序之后用双指针
'''
两个指针遍历排序好的数组
若是指的元素相同，添加到新的数组
不同
值小的指针后移

'''

def intersect(nums1,nums2):
    """
    :type nums1:List[int]
    :type nums2:List[int]
    :rtype List[int]
    """
    nums1.sort()
    nums2.sort()
    nums = []
    i,j = 0,0
    while i < len(nums1) and j < len(nums2):
        if nums1[i] == nums2[j]:
            nums.append(nums1[i])
            i += 1
            j += 1
        elif nums1[i] < nums2[j]:
            i += 1
        else:
            j += 1
    return nums



if __name__ == "__main__":
    #nums1 = [4,9,5,5,4]
    #nums2=[9,4,9,8,4,9,8,4,9]
    nums1 = [1,2,2,1]
    nums2 = [2,2] 
    print(intersect(nums1,nums2))

File: xiaohao/数组/350/350_双指针解法.go
package main

import (
	"fmt"
	"sort"
)

func insert(nums1 []int, nums2 []int) []int {
	ptr1, ptr2, length, k := 0, 0, 0, 0

	//排序
	sort.Ints(nums1)
	sort.Ints(nums2)
	fmt.Println(nums1)
	fmt.Println(nums2)

	if len(nums1) < len(nums2) {
		length = len(nums1)
	} else {
		length = len(nums2)
	}

	nums := make([]int, length)

	for ptr1 < len(nums1) && ptr2 < len(nums2) {
		if nums1[ptr1] == nums2[ptr2] {
			nums[k] = nums1[ptr1]
			ptr1 += 1
			ptr2 += 1
			k += 1
		} else if nums1[ptr1] < nums2[ptr2] {
			ptr1+=1
		} else {
			ptr2+=1
		}
	}
	return nums[:k]
}

func main() {
	nums1 := []int{11, 1, 11, 2, 4, 7, 3, 11, 9, 12, 5, 7, 7, 6}
	nums2 := []int{6, 6, 6, 6, 7, 7, 10, 11, 11}
	fmt.Println(insert(nums1, nums2))

}
